tracer/src/Datadog.Trace/Iast/IastModule.cs

```diff
@@ -41,102 +43,166 @@ static IastModule()
     }
 
-    internal static Scope? OnTrustBoundaryViolation(string name)
+    internal static string OnUnvalidatedRedirect(string evidence)
+    {
+        if (OnUnvalidatedRedirect(evidence, IntegrationId.UnvalidatedRedirect).VulnerabilityAdded)
+        {
+            return new string(evidence.ToCharArray());
+        }
+
+        return evidence;
+    }
+
+    internal static IastModuleResponse OnUnvalidatedRedirect(string evidence, IntegrationId integrationId)
+    {
+        bool HasInvalidOrigin(TaintedObject tainted)
+        {
+            try
+            {
+                foreach (var range in tainted.Ranges)
+                {
+                    if (range.Source is null) { continue; }
+                    var origin = (SourceTypeName)range.Source.OriginByte;
+                    // TODO: reenable when SourceTypeName.Database gets defined -> if (origin == SourceTypeName.Database) { continue; }
+                    if (origin == SourceTypeName.RequestPath) { continue; }
+                    if (origin == SourceTypeName.RequestHeaderValue && range.Source.Name == "Host") { continue; }
+
+                    return true;
+                }
+
+                return IsRefererHeader(tainted);
+            }
+            catch (Exception err)
+            {
+                Log.Error(err, "Error while checking for UnvalidatedRedirect tainted origins.");
+            }
+
+            return true;
+        }
+
+        bool IsRefererHeader(TaintedObject tainted)
+        {
+            foreach (var range in tainted.Ranges)
+            {
+                if (range.Source is null) { return false; }
+                var origin = (SourceTypeName)range.Source.OriginByte;
+                if (origin != SourceTypeName.RequestHeaderValue || !ReferrerHeaderName.Equals(range.Source.Name, StringComparison.OrdinalIgnoreCase))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        try
+        {
+            OnExecutedSinkTelemetry(IastInstrumentedSinks.UnvalidatedRedirect);
+            return GetScope(evidence, integrationId, VulnerabilityTypeName.UnvalidatedRedirect, OperationNameUnvalidatedRedirect, HasInvalidOrigin);
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Error while checking for Unvalidated Redirect.");
+            return IastModuleResponse.Empty;
+        }
+    }
+
+    internal static IastModuleResponse OnTrustBoundaryViolation(string name)
     {
         try
         {
             OnExecutedSinkTelemetry(IastInstrumentedSinks.TrustBoundaryViolation);
-            return GetScope(name, IntegrationId.TrustBoundaryViolation, VulnerabilityTypeName.TrustBoundaryViolation, OperationNameTrustBoundaryViolation, true);
+            return GetScope(name, IntegrationId.TrustBoundaryViolation, VulnerabilityTypeName.TrustBoundaryViolation, OperationNameTrustBoundaryViolation, (t) => true);
```

Comment:
I would imagine `(t) => true` is allocated as a static lambda and only allocated one time since it doesn't capture any state, but maybe we could have a static method called `Always` to make it slightly more readable. What do you think?

Reply:
Good idea. Implementing it